### PR TITLE
Fix "Cannot read property 'attributes' of null"

### LINKED ISSUE #698: Uncaught TypeError: Cannot read property 'attributes' of null
## Are you reporting a feature or a bug?

Bug

## Provide detailed reproduction steps (if any)

In some very specific cases, editable inline widgets will throw an "_Uncaught TypeError: Cannot read property 'attributes' of null_" after making a modification to the content of the widget. 

Please follow the steps below using [this JSFiddle](https://jsfiddle.net/a6xrqx2u/):

1. Select the very last letter of the text inside the widget (= select the "m" of "lorem")
2. Click the **B**old button from the toolbar
3. Switch to Source mode

### Actual result

The following error occurs:
```
Uncaught TypeError: Cannot read property 'attributes' of null
    at VM232 plugin.js?t=H5SB:23
    at window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.htmlParser.element.forEach (VM226 ckeditor.js:281)
    at window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.htmlParser.element.forEach (VM226 ckeditor.js:281)
    at window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.htmlParser.element.forEach (VM226 ckeditor.js:281)
    at a.<anonymous> (VM232 plugin.js?t=H5SB:23)
    at a.f (VM226 ckeditor.js:11)
    at a.window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.event.CKEDITOR.event.fire (VM226 ckeditor.js:12)
    at a.window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.editor.CKEDITOR.editor.fire (VM226 ckeditor.js:14)
    at window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.htmlDataProcessor.toDataFormat (VM226 ckeditor.js:302)
    at $.getData (VM226 ckeditor.js:1183)

```

## Other details

* Browser: Chrome 59.0.3071.115 (Official Build) (64-bit)
* OS: Windows 10 (Creator Update)
* CKEditor version: 4.7.1
* Installed CKEditor plugins: "full-all" preset

### PR DESCRIPTION
**Pull request by glen-84**
## Does your PR contain necessary tests?

No, it's a minor change. Let me know if this is necessary.

### This PR contains

- [ ] Unit tests
- [ ] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#698](https://github.com/ckeditor/ckeditor4/issues/698): Fix "Cannot read property 'attributes' of null".
```

## What changes did you make?

Checked that `element` and `widgetElement` are non-null before accessing `attributes`.

## Which issues does your PR resolve?

Closes #698.

### CODE REVIEW

**Review by Comandeer — CHANGES_REQUESTED** [review]
Hi, @glen-84! Thanks for your contribution.

As all changes to the CKEditor 4 codebase require tests, can I ask you to add some? We have a [guide describing our test environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html).

Additionally, the fix does not seem to resolve all issues around switching to the source mode. The original issue is fixed but when I tried the [original reproduction](https://jsfiddle.net/a6xrqx2u/) with your fix, I got another error:

```javascript
node.js:45 Uncaught TypeError: Cannot set properties of null (setting 'previous')
    at CKEDITOR.htmlParser.element.replaceWith (node.js:45:30)
    at Editor.<anonymous> (plugin.js:3119:18)
    at Editor.listenerFirer (event.js:169:34)
    at CKEDITOR.event.fire (event.js:313:36)
    at CKEDITOR.editor.fire (editor_basic.js:24:40)
    at CKEDITOR.htmlDataProcessor.toDataFormat (htmldataprocessor.js:301:23)
    at $.getData (plugin.js:623:34)
    at $.<anonymous> (editable.js:839:22)
    at Editor.listenerFirer (event.js:169:34)
    at CKEDITOR.event.fire (event.js:313:36)
```

Could you check if it can be easily fixed?

**Comment by glen-84** [comment]
@Comandeer

I wrote this test:

```js
/* bender-tags: widget */

( function() {
	'use strict';

	CKEDITOR.dtd.$editable.span = 1;

	CKEDITOR.plugins.add( 'spanwidget', {
		requires: 'widget',
		init: function ( editor )	{
			editor.widgets.add( 'spanwidget', {
				editables: {
					content: {
						selector: 'span'
					}
				},
				upcast: function ( element ) {
					return element.name == 'span';
				}
			});
		}
	});

	bender.editor = {
		creator: 'replace',
		config: {
			allowedContent: true,
			extraPlugins: 'basicstyles,embed,sourcearea,spanwidget'
		}
	};

	bender.test( {
		'Fix #698': function() {
			var tc = this;

			tc.editor.setData( '<strong><span>lorem</span></strong>', function() {
				tc.resume( function() {
					tc.editor.focus();

					var textNode = tc.editor.editable().findOne( 'span' ).getFirst().getFirst();

					// Select the `m`.
					var rng = tc.editor.createRange();
					rng.setStart( textNode, 4 );
					rng.setEnd( textNode, 5 );
					tc.editor.getSelection().selectRanges( [ rng ] );

					tc.editor.execCommand( 'bold' )

					tc.editor.setMode( 'source' );

					assert.isTrue( true ); // Assert no errors?
				}, 100 );
			} );

			tc.wait();
		}
	} );

} )();
```

But it triggers the error that you mentioned, and if I add `if ( !widgetElement ) continue;` in `setupWidgetsLifecycleEnd`, the source editor shows HTML from CKE:

```html
<p><strong><span>lorem</span></strong><strong><span tabindex="-1" contenteditable="false" role="region" aria-label="span widget"><span style="background:rgba(220,220,220,0.5);background-image:url(http://localhost:1030/apps/ckeditor/plugins/widget/images/handle.png);display:none;"><img src="data:image/gif;base64,R0lGODlhAQABAPABAP///wAAACH5BAEKAAAALAAAAAABAAEAAAICRAEAOw==" width="15" title="Click and drag to move" height="15" role="presentation" draggable="true" /></span></span></strong></p>
```

I'm really out of my depth here. If you're able to assist with finishing this up, I'd appreciate that.

**Comment by Comandeer** [comment]
@glen-84, I've added your test (slightly modified to better fit with the style with our newer tests but the overall logic is left untouched) alongside a manual one.

The `setupWidgetsLifecycleEnd()` was a correct place to apply a fix. Although it's probably better to [simply remove the widget](eb14c12b4b09a682759a1076618f26c3946b591167ad6cbed255dR3119-R3128) if there is no `retElement` at the end of this function. Especially as applying bold to the last letter in the `span` causes the widget to be duplicated.

I see that you also added an additional check to the `embedsemantic` plugin. Could you explain why is it needed (or describe a case when the issue occurs in that plugin)?

**Comment by Comandeer** [comment]
> My plugin uses a span, not a div.

That's probably the reason why the issue appears in your plugin. There is no official support for inline editables in CKEditor 4 so some things could work incorrectly.

**Review by Comandeer — APPROVED** [review]
LGTM! Thanks for your contribution, @glen-84!

However, as I've also added some changes to the PR, I'll ask @KarolDawidziuk to have a second look at it.

File: tests/plugins/widget/manual/inlineeditable.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+1. Switch to the source mode.
+
+	**Expected** Editor is switched to the source mode.
+
+	**Unexpected** There is an error in the console and editor is not switched to the source mode.
+1. Switch back to the editing mode.
+
+	**Expected** Editor is switched to the editing mode.
+
+	**Unexpected** There is an error in the console and editor is not switched to the editing mode.
```
Comment (KarolDawidziuk) [inline tests/plugins/widget/manual/inlineeditable.md:17]: Based on our [code style](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md#31-manual-tests) this additional indentation is unnecessary, let's remove this.
```suggestion
**Expected** Editor is switched to the source mode.

**Unexpected** There is an error in the console and editor is not switched to the source mode.

1. Switch back to the editing mode.

**Expected** Editor is switched to the editing mode.

**Unexpected** There is an error in the console and editor is not switched to the editing mode.
```

File: tests/plugins/widget/manual/inlineeditable.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+@bender-tags: widget, bug, 698, 4.21.0
+@bender-ui: collapsed
+@bender-ckeditor-plugins: wysiwygarea, toolbar, sourcearea, basicstyles, widget
+
+1. Open browser console.
+1. Select `t` (the last letter) from `Test`.
+1. Click the "Bold" toolbar button.
```
Comment (KarolDawidziuk) [inline tests/plugins/widget/manual/inlineeditable.md:7]:
```suggestion
1. Click the `Bold` toolbar button.
```

File: tests/plugins/widget/manual/inlineeditable.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
+@bender-tags: widget, bug, 698, 4.21.0
```
Comment (KarolDawidziuk) [inline tests/plugins/widget/manual/inlineeditable.md:1]: Based on our [code style](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md#31-manual-tests) let's use a correct tag order:
```suggestion
@bender-tags: 4.21.0, bug, 698, widget
```

**Review by KarolDawidziuk — CHANGES_REQUESTED** [review]
Thanks for your improvements.

The main issue with throwing an error is solved, good job! 
There are some code-style things(mostly nitpicks) that we can improve. Please see my inline comments.

Also, there is an issue with inserting an additional empty `<strong>` element in the iframe-based editor when we switch to the source mode, but this is not in the scope of this PR.
Let's extract this as a follow-up and after you apply changes from inline comments PR will be ready to merge.

File: plugins/widget/plugin.js [hunk 4]
```diff
@@ -3059,7 +3059,7 @@
 						} );
 
 						// If widget did not have data-cke-widget attribute before upcasting remove it.
-						if ( widgetElement.attributes[ 'data-cke-widget-keep-attr' ] != '1' )
+						if ( widgetElement && widgetElement.attributes[ 'data-cke-widget-keep-attr' ] != '1' )
 							delete widgetElement.attributes[ 'data-widget' ];
```
[the remark this answers is not in the document]
Comment (glen-84) [reply plugins/widget/plugin.js:3063]: This is existing code, unrelated to this PR. 🙂

**Comment by glen-84** [comment]
I'll leave the other suggestions for @Comandeer, since they are related to his changes.

**Comment by Comandeer** [comment]
@KarolDawidziuk, thx for the review. I've applied your suggestions and PR is ready for another one.

**Review by KarolDawidziuk — APPROVED** [review]
LGTM! 🎉

**Comment by Comandeer** [comment]
Merged. @glen-84, thanks once more for your contribution!

**Pull request merged**